collectors/debugfs.plugin/sys_devices_virtual_powercap.c

```diff
@@ -0,0 +1,246 @@
+}
+
+static double
+calculate_watts(unsigned long long max_energy_range_uj, struct measurement_t *before, struct measurement_t *after)
+{
+    unsigned long long energy_uj = 0;
+    usec_t delta_us = after->time_us - before->time_us;
+    if (unlikely(delta_us == 0)) {
+        return 0;
+    }
+
+    if (likely(after->energy_uj >= before->energy_uj)) {
+        energy_uj = after->energy_uj - before->energy_uj;
+    } else {
+        energy_uj = after->energy_uj + (max_energy_range_uj - before->energy_uj);
+    }
+    return (double)energy_uj / delta_us;
+}
+
+static struct zone_t *
+get_zone(const char *control_type, struct zone_t *parent, const char *dirname, struct zone_t **zones, size_t *count)
+{
+    char temp[FILENAME_MAX + 1];
+    snprintfz(temp, FILENAME_MAX, "%s/%s", dirname, "name");
+
+    char name[FILENAME_MAX + 1] = "";
+    if (read_file(temp, name, sizeof(name) - 1) != 0) {
+        return NULL;
+    }
+
+    char *trimmed = trim(name);
+    if (unlikely(trimmed == NULL || trimmed[0] == 0)) {
+        return NULL;
+    }
+
+    snprintfz(temp, FILENAME_MAX, "%s/%s", dirname, "max_energy_range_uj");
+    unsigned long long max_energy_range_uj = 0;
+    if (unlikely(read_single_number_file(temp, &max_energy_range_uj) != 0)) {
+        collector_error("Cannot read %s", temp);
+        return NULL;
+    }
+
+    snprintfz(temp, FILENAME_MAX, "%s/%s", dirname, "energy_uj");
+    struct measurement_t measurement;
+    if (unlikely(get_measurement(temp, &measurement) != 0)) {
+        collector_error("%s: Cannot read %s", trimmed, temp);
+        return NULL;
+    }
+
+    int parent_idx = -1;
+    if (parent != NULL) {
+        parent_idx = parent - *zones;
+    }
+
+    *zones = (struct zone_t *)reallocz(*zones, sizeof(struct zone_t) * (*count + 1));
+    if (unlikely(*zones == NULL)) {
+        collector_error("reallocz failed");
+        *count = 0;
+        return NULL;
+    }
+
+    struct zone_t *zone = &(*zones)[*count];
+    strcpy(zone->path, temp);
+    if (parent_idx >= 0) {
+        snprintfz(zone->name, FILENAME_MAX, "%s/%s", (*zones)[parent_idx].name, trimmed);
+    } else {
+        snprintfz(zone->name, FILENAME_MAX, "%s/%s", control_type, trimmed);
+    }
+    zone->max_energy_range_uj = max_energy_range_uj;
+    zone->measurement = measurement;
+
+    collector_info("Found zone: \"%s\"", zone->name);
+
+    ++(*count);
+
+    return zone;
+}
+
+static void
+look_for_zones(const char *control_type, struct zone_t *parent, const char *path, struct zone_t **zones, size_t *count)
+{
+    DIR *dir = opendir(path);
+    if (unlikely(dir == NULL)) {
+        return;
+    }
+
+    struct dirent *de = NULL;
+    while ((de = readdir(dir))) {
+        if (de->d_type != DT_DIR || de->d_name[0] == '.') {
+            continue;
+        }
+
+        char zone_path[FILENAME_MAX + 1];
+        snprintfz(zone_path, FILENAME_MAX, "%s/%s", path, de->d_name);
+
+        collector_info("Looking for zone in \"%s\"", zone_path);
```

Comment:
Is there a specific reason, that is not clear for me, for users have logs every time we read?  I liked the idea to have these tests behind some `macro` like `NETDATA_DEV_MODE` that we use to develop our code.

Reply:
These logs only occur once during netdata startup. As a user I find it useful in cases where I was expecting to see some metric but it's not showing up. Please let me know if you still want to me only log it in dev mode.

Reply:
I got your point, thank you very much for your contribution @fhriley , please, take a look in @ilyam8's comment.